Tests and testing including smoke tests

wendellpiez opened the pull request:
# Committer Notes

To address #53

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/metaschema-xslt/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/metaschema-xslt/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [ ] Do all automated CI/CD checks pass?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)?
- [x] Have you updated all [website](https://pages.nist.gov/metaschema) and readme documentation affected by the changes you made? Changes to the website can be made in the website/content directory of your branch.

wendellpiez committed:
Basic XSD testing - validating and proofing an XSD generated for a metaschema

wendellpiez committed:
Makefile for converter generator

wendellpiez committed:
Added xsd-proofcheck script to metaschema-xslt make script setup

wendellpiez committed:
Don't forget to update help text

wendellpiez committed:
More readme and script touchups

wendellpiez committed:
Merge branch 'develop' into issue53-smoketesting

aj-stein-nist commented on src/schema-gen/mvn-xsd-proofcheck-xpl.sh:
```diff
@@ -0,0 +1,33 @@
+CALABASH_ARGS1="-iMETASCHEMA=\"$METASCHEMA_XML\" -oOUT_XSD=/dev/null \
+                $ADDITIONAL_ARGS \"${SCRIPT_DIR}/METASCHEMA-XSD-TEST.xpl\""
+
+invoke_calabash "${CALABASH_ARGS0}"
+invoke_calabash "${CALABASH_ARGS1}"
```
Blocking: as a consumer of such tools, I would prefer some error checking based on exit codes with `result=$(invoke_calabash "${CALABASH_ARGS0}")` and processing the all is well message or numeric code so I can layer on top of it.

wendellpiez replied:
I would love to see how this is done! (Up for a demo?)

nikitawootten-nist replied:
Currently all subcommands source [`src/common/subcommand_common.bash`](https://github.com/usnistgov/metaschema-xslt/blob/develop/src/common/subcommand_common.bash), which among other things, sets `-Eeuo pipefail` which will enforce the script exiting immediately if an error occurs in either of the `invoke_calabash` calls.

aj-stein-nist commented on src/README.md:
````diff
@@ -24,34 +24,6 @@ The help message includes a list of the supported subcommands, indicating which
 
 See each subdirectory README for more instructions.
 
-#### Dedicated scripts
-
-Within any of the subdirectories in `src`, recognize the scripts by their `.sh` file suffix. The scripts follow a naming convention, with an initial segment identifying the primary executable invoked by the script (usually `mvn` for Maven); a final segment `xpl` or `xsl` indicating XPoc or XSLT entries, and intermediate segments indicating what the script produces.
-
-For example, `mvn-xsd-schema-xsl.sh` can be run to produce an XSD schema from a metaschema, using an XSLT-based process (i.e., Saxon with an appropriate XSLT transformation), run under Maven.
-
-Each script also requires arguments, typically the path to the metaschema source (input) file along with a name or keyword directing where to write results. Invoke the script without arguments to get help on its syntax requirements.
-
-Scripts and stylesheets are also documented in place using readmes and in line. Since XSLTs can call, import, include or read XSLTs from elsewhere in the distribution, and sometimes do, keep the modules together: each folder on its own is *not* self-contained.
-
-Users may also apply and use resources in this repository in their own scripts. Do this either by cloning, copying and modifying scripts here; by writing your own scripts or shells; or by adapting code into the XML/XSLT processing framework or stack of your choice.
-
-A convention is used indicating that an XProc (`*.xpl` file) or XSLT (`*.xsl`) intended to be invoked directly (that is, not only to be used as a module or component) is given a name entirely or partly in `ALL-CAPITALS`. For example, `src/schema-gen/METASCHEMA-ALL-SCHEMAS.xpl` is such an XProc pipeline (a step definition intended to be used directly). The XSLTs that observe this convention are, a ...
-
-### `make` support
-
-Additionally, some subdirectories include `make` configurations. These are used for testing including regression testing, but may also be used to support processing.
-
-To use `make`, confirm you have [`make`](https://www.gnu.org/software/make/), or install it. In any directory with a Makefile, including this one, test it:
-
-```
-> src/schema-gen make
-```
-
-The system returns a list of available (configured) targets, typically running tests.
-
-## Subdirectories
-
````
Blocking: did you intend to delete this and not replace it? I just want to check, not trying to read too much into it.

wendellpiez replied:
Argh, good catch. I thought that stuff was moved but apparently not. How do I revoke this commit?

wendellpiez replied:
Okay I think I just patched it back in. There is some redundancy across the top-level and `src` readmes, but nothing I think I can't live with.

aj-stein-nist reviewed (commented):
I do not enough time this morning to do a complete, proper review, but here is some preliminary feedback since I got the ping to review it. :-)

wendellpiez committed:
Restored accidental removal

david-waltermire commented on src/schema-gen/XSD/schema/XMLSchema-1.0.xsd:
```diff
@@ -0,0 +1,2534 @@
+      towards the end of this schema document</xs:documentation>
+ </xs:annotation>
+
+ <xs:import namespace="http://www.w3.org/XML/1998/namespace" schemaLocation="http://www.w3.org/2001/xml.xsd">
```
Blocking: This schema imports other schema, which should also be saved locally.

Non-blocking: These schema are available in the metaschema submodule. You could use these instead of making an additional copy.

wendellpiez replied:
Excellent points @david-waltermire-nist. For now I have uploaded xml.xsd.

To the second point -- the XSD is an XSD 1.0 XSD for XSD, such as can be supported by XML Calabash, which AFAIK isn't able to support validation under 1.1 (so we do what we can). Consolidating schema directories would mean keeping both 1.0 and 1.1 XSDs for XSD in `metaschema` repo. I can make an Issue for that but only if you say you want it.

david-waltermire commented on src/schema-gen/TESTING.md:
```diff
@@ -0,0 +1,27 @@
+
+- XSD
+- JSON Schema, v7
+- (planned) Schematron
```
Consider documenting plans as issues, not in documentation. IMHO, documentation should explain what currently exists.

wendellpiez replied:
Removed from latest.

david-waltermire reviewed (changes requested):
I believe some changes are needed to tighten up this PR. See associated comments.

wendellpiez committed:
Update TESTING.md

Tried to generalize the description away from current state of coverage (wrt schema languages)

wendellpiez committed:
Adding xml.xsd (dependency for included XSD 1.0 XSD)

wendellpiez commented:
@aj-stein-nist let's make a time to confer re this PR. Not only b/c I need you to help unblock it, but because I am seeing a couple of files on my system that *might* belong in this PR - has anyone but me run it yet? thanks!

aj-stein-nist commented:
> @aj-stein-nist let's make a time to confer re this PR. Not only b/c I need you to help unblock it, but because I am seeing a couple of files on my system that _might_ belong in this PR - has anyone but me run it yet? thanks!

Sorry I dropped the ball on this. Per discussion this afternoon I will take a look and double check, confirming with a review if things seem out of alignment or not.

aj-stein-nist commented on src/schema-gen/mvn-xsd-proofcheck-xpl.sh:
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
```
Minor, but blocking: Can you please mark this executable by running this from WSL2 on your computer? This makes it much easier to use for others and ready to bake in automated pipelines.

```sh
chmod +x ./src/schema-gen/mvn-xsd-proofcheck-xpl.sh
```

aj-stein-nist replied:
I was able to push it myself, so you can disregard this pending review item and I will approve.

aj-stein-nist reviewed (changes requested):
Can we please make use of the metaschema submodule resources and not reduplicate the relevant files in `XSD/schema`? Also, make the new script executable. Otherwise, this looks good.

wendellpiez committed:
Added clarifying readme.md for included schemas

wendellpiez commented:
@aj-stein-nist please check my work?

- chmod +x on .sh file (and I need to see to this being done by default)
- readme added to explain not-actually-superfluous schema files

aj-stein-nist commented:
>     * chmod +x on .sh file (and I need to see to this being done by default)

```sh
$ git log
commit 03715db7b85c7883904133c3cdd6b3a1b064f1f5 (HEAD -> issue53-smoketesting, wendell/issue53-smoketesting)
Author: Wendell Piez <wendell.piez@nist.gov>
Date:   Mon Dec 4 16:49:33 2023 -0500

    Added clarifying readme.md for included schemas
$ git rev-parse HEAD
03715db7b85c7883904133c3cdd6b3a1b064f1f5
$ ls -lha ./src/schema-gen/mvn-xsd-proofcheck-xpl.sh
-rw-r--r-- 1 me me 1.2K Nov 30 12:00 ./src/schema-gen/mvn-xsd-proofcheck-xpl.sh # The lack of x means still not executable
$ ./src/schema-gen/mvn-xsd-proofcheck-xpl.sh
bash: ./src/schema-gen/mvn-xsd-proofcheck-xpl.sh: Permission denied
```

I still do not see it working on my end after pulling down changes. I guess we sync up and take a look in a few minutes if you are available?

>     * readme added to explain not-actually-superfluous schema files

That is more than sufficient, you could have told me in a comment and I would have walked back that comment. Thanks for the nice explanation.

aj-stein-nist committed:
Make mvn-xsd-proofcheck-xpl.sh executable

Use `chmod +x` to make the aforementioned script executable and resolve
my pending PR review feedback.

wendellpiez commented:
@aj-stein-nist could I trouble you to try again?

After confirming the permissions on my files, I ran

```
 git update-index --chmod +x *.sh
```

I then pushed this back up. Does it help? Grateful as always for the assistance.
